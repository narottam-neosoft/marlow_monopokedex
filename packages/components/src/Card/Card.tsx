import React, { useState } from "react";
import MuiCard from "@mui/material/Card";
import {
  Box,
  CardContent,
  CardMedia,
  CircularProgress,
  List,
  ListItem,
  ListItemText,
  Typography,
} from "@mui/material";
import { cardProps, IAbilityProp } from "./Card.types";

export const Card = (props: cardProps) => {
  const [imageLoaded, setImageLoaded] = useState(false);

  const { image, name, height, weight, location, abilities, stats, sprites } =
    props;

  return (
    <MuiCard
      sx={{
        display: "flex",
        boxShadow: "rgba(0, 0, 0, 0.16) 0px 10px 36px 0px, rgba(0, 0, 0, 0.06) 0px 0px 0px 1px",
      }}
    >
      <CardContent>
        <Box
          style={{
            display: "flex",
            flexDirection: "column",
            justifyContent: "center",
            alignItems: "center",
          }}
        >
          {!imageLoaded && (
            <Box
              sx={{
                width: "50%",
                display: "flex",
                alignItems: "center",
                justifyContent: "center",
              }}
            >
              <CircularProgress />
            </Box>
          )}
          <Box sx={{ display: "flex", alignItems: "center", gap: "5px" }}>
            {imageLoaded && (
              <CardMedia
                component="img"
                alt="front_image"
                height="auto"
                width="auto"
                image={sprites.front_default}
              />
            )}
            <img
              src={image}
              alt="pokedex character"
              style={{ display: "none" }}
              onLoad={() => setImageLoaded(true)}
              onError={() => setImageLoaded(true)}
            />
            <CardMedia
              sx={{
                width: "150px",
                height: `${imageLoaded}? "150px" : "0px" `,
                objectFit: "contain",
              }}
              component="img"
              image={image}
              title={name}
              loading="lazy"
              style={{ visibility: imageLoaded ? "visible" : "hidden" }}
            />
            {imageLoaded && (
              <CardMedia
                component="img"
                alt="front_image"
                height="auto"
                width="auto"
                image={sprites.back_default}
              />
            )}
          </Box>
          <Typography
            gutterBottom
            variant="h4"
            component="div"
            sx={{ textTransform: "capitalize" }}
          >
            {name}
          </Typography>
        </Box>

        <Box sx={{ display: "flex", gap: 5 }}>
          <Typography variant="body2" color="text.secondary">
            Height : {height}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            Weight : {weight}
          </Typography>
        </Box>
        <Typography
          variant="body2"
          color="text.secondary"
          sx={{ display: "flex", flexDiretion: "column" }}
        >
          Location:
          <a href={location} target="_blank" rel="noreferrer">
            {location}
          </a>
        </Typography>
        <Box>
          <Typography variant="h6" component="div">
            Abilities
          </Typography>
          <List dense={false} sx={{ lineHeight: "1px" }}>
            {abilities.map((ability: IAbilityProp, i: number) => (
              <ListItem key={i}>
                <ListItemText
                  primary={ability.ability.name}
                  secondary={ability.ability.url}
                />
              </ListItem>
            ))}
          </List>
        </Box>
        <Box>
          <Typography variant="h6" component="div">
            Stats
          </Typography>
          <Box sx={{ pl: "16px" }}>
            {stats?.map((statDetails, index) => (
              <Box key={index}>
                <Typography
                  sx={{ display: "inline" }}
                  component="div"
                  variant="subtitle1"
                >
                  {statDetails?.stat?.name} :{" "}
                </Typography>
                <Typography
                  sx={{ display: "inline" }}
                  component="div"
                  variant="subtitle2"
                >
                  {statDetails?.base_stat}
                </Typography>
              </Box>
            ))}
          </Box>
        </Box>
      </CardContent>
    </MuiCard>
  );
};
