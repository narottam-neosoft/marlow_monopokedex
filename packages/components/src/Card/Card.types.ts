export type IAbilityProp = {
  ability: {
    name: string;
    url: string;
  };
};
export interface cardProps {
  name: string;
  height?: number;
  weight?: number;
  location?: string;
  abilities?: IAbilityProp[];
  image?:string;
  stats?: {
    base_stat: number;
    stat: {
      name: string;
    }
  }[];
  sprites: {
    front_default: string;
    back_default: string;
  }
}
