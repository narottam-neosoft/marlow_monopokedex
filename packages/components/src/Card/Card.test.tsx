import * as React from "react";
import { render, screen } from "@testing-library/react";
import { Card } from "./Card";


describe("Card", () => {
  const mockProps = {
    name: "Pikachu",
    image: "pikachu.jpg",
    height: 23,
    weight: 34,
    location: "https://pokeapi.co/api/v2/pokemon/25",
    abilities: [
      {
        ability: {
          name: "Ability 1",
          url: "https://pokeapi.co/api/v2/ability/1",
        },
      },
      {
        ability: {
          name: "Ability 2",
          url: "https://pokeapi.co/api/v2/ability/2",
        },
      },
    ],
  };

  it("renders the card with correct props", () => {
    const { container } = render(<Card {...mockProps} />);

    expect(screen.getByText("Pikachu")).toBeInTheDocument();
    const locationLink = screen.getByRole("link", {
      name: "https://pokeapi.co/api/v2/pokemon/25",
    });
    expect(locationLink).toBeInTheDocument();
    expect(locationLink).toHaveAttribute(
      "href",
      "https://pokeapi.co/api/v2/pokemon/25"
    );
    expect(container).toMatchSnapshot();
  });
});
