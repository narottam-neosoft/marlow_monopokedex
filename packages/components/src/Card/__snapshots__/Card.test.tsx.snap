// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders the card with correct props 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1oqv2c0-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-vnfzf1"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
    <img
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-l4hmoc-MuiCardMedia-root"
      loading="lazy"
      src="pikachu.jpg"
      title="Pikachu"
    />
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-3ra5la-MuiTypography-root"
      >
        Pikachu
      </div>
      <div
        class="MuiBox-root css-16wajk8"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Height : 
          23
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Weight : 
          34
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-1q3l2w3-MuiTypography-root"
      >
        Location:
        <a
          href="https://pokeapi.co/api/v2/pokemon/25"
          rel="noreffer"
          target="_blank"
        >
          https://pokeapi.co/api/v2/pokemon/25
        </a>
      </p>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          Abilities
        </div>
        <ul
          class="MuiList-root MuiList-padding css-1ept1ed-MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Ability 1
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                https://pokeapi.co/api/v2/ability/1
              </p>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Ability 2
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                https://pokeapi.co/api/v2/ability/2
              </p>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
